File: Leetcode/72 Edit distance/minDistance.cpp
/*This code is contributed by mr-optimizer (https://github.com/mr-optimizer) */
class Solution {
public:
    int memo[501][501];
    int dp(string word1, string word2, int i, int j) {
        if (i == 0)return j;
        if (j == 0)return i;
        if (memo[i][j] != -1)return memo[i][j];
        if (word1[i - 1] == word2[j - 1]) {
            return memo[i][j] = dp(word1, word2, i - 1, j - 1);
        } else {
            return memo[i][j] = 1 + min(dp(word1, word2, i - 1, j), min(dp(word1, word2, i, j - 1), dp(word1, word2, i - 1, j - 1)));
        }
    }
    int minDistance(string word1, string word2) {
        memset(memo, -1, sizeof(memo));
        return dp(word1, word2, word1.size(), word2.size());
    }
};

File: Leetcode/72 Edit distance/minDistance.java
public class Solution {
    public int minDistance(String A, String B) {
        int[][]list=new int[A.length()][B.length()];
        return editDist(A,B,0,0,list);
    }
    int editDist(String A,String B,int x,int y,int[][]memory){
        if(y==B.length())return A.length()-x;
        if(x==A.length())return B.length()-y;
        int ans=memory[x][y];
        if(ans!=0)return ans;
        if(A.charAt(x)==B.charAt(y)){
            ans=editDist(A,B,x+1,y+1,memory);
        }else{
            ans=1+Math.min(editDist(A,B,x+1,y,memory),Math.min(editDist(A,B,x,y+1,memory),editDist(A,B,x+1,y+1,memory)));
        }
        memory[x][y]=ans;
        return ans;
    }
}
